### ISSUE TITLE
RPC test failures (BadStatusLine / No such file or directory: 'cache/node1') with recent Python, e.g. Python 2.7.16

### ISSUE THREAD

**Issue by daira** [zcash/zcash#4087]
### Describe the issue
RPC tests always fail with errors like the one below.

#### If so, please list the steps to reproduce below:
1. Dependencies already installed and zcutil/fetch-params.sh has been run.
2. git clone https://github.com/zcash/zcash && cd zcash && zcutil/build.sh -j12
3. qa/pull-tester/rpc-tests.sh

### Expected behaviour
RPC tests should all pass.

### Actual behaviour + errors
I cannot get RPC tests to work at all on my machine, even in a new checkout with all deps compiled from scratch.

```$ qa/pull-tester/rpc-tests.sh 
=== Running testscript paymentdisclosure.py ===
Initializing test directory /tmp/testnx2xkV
Mining blocks...
waiting for async operation opid-1510dac9-99ca-456a-b33c-7a178cb29d2b
Unexpected exception caught during testing: No status line received - the server has closed the connection
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/test_framework.py", line 121, in main
    self.run_test()
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/paymentdisclosure.py", line 80, in run_test
    self.sync_all()
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/test_framework.py", line 76, in sync_all
    sync_blocks(self.nodes)
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/util.py", line 52, in sync_blocks
    counts = [ x.getblockcount() for x in rpc_connections ]
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/authproxy.py", line 143, in __call__
    response = self._request('POST', self.__url.path, postdata)
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/authproxy.py", line 120, in _request
    return self._get_response()
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/authproxy.py", line 158, in _get_response
    http_response = self.__conn.getresponse()
  File "/usr/lib/python2.7/httplib.py", line 1121, in getresponse
    response.begin()
  File "/usr/lib/python2.7/httplib.py", line 438, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python2.7/httplib.py", line 402, in _read_status
    raise BadStatusLine("No status line received - the server has closed the connection")
Stopping nodes
Traceback (most recent call last):
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/paymentdisclosure.py", line 218, in <module>
    PaymentDisclosureTest().main()
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/test_framework.py", line 137, in main
    stop_nodes(self.nodes)
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/util.py", line 227, in stop_nodes
    node.stop()
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/authproxy.py", line 143, in __call__
    response = self._request('POST', self.__url.path, postdata)
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/authproxy.py", line 119, in _request
    self.__conn.request(method, path, postdata, headers)
  File "/usr/lib/python2.7/httplib.py", line 1042, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python2.7/httplib.py", line 1076, in _send_request
    self.putrequest(method, url, **skips)
  File "/usr/lib/python2.7/httplib.py", line 924, in putrequest
    raise CannotSendRequest()
httplib.CannotSendRequest
!!! FAIL: paymentdisclosure.py !!!
```

That's the first test. They all fail with similar errors.

`zcashd` itself seems to work fine and syncs successfully on this machine.

### The version of Zcash you were using:
```
$ src/zcashd --version
Zcash Daemon version v2.0.6-12046373b
```

Same symptoms with the v2.0.6 release.

### Machine specs:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID:    Debian
Description:    Debian GNU/Linux 10 (buster)
Release:    10
Codename:    buster
$ uname -a
Linux katava 4.19.0-5-amd64 #1 SMP Debian 4.19.37-5 (2019-06-19) x86_64 GNU/Linux
```

- CPU: Intel(R) Core(TM) i7-6800K CPU @ 3.40GHz
- RAM: 96 GiB (45 GiB free)
- Disk size: 420 GB free
- Disk Type (HD/SDD): SDD

```
$ python --version
Python 2.7.16
$ gcc --version
gcc (Debian 8.3.0-6) 8.3.0
$ ld -v
GNU ld (GNU Binutils for Debian) 2.31.1
$ as --version
GNU assembler (GNU Binutils for Debian) 2.31.1
```

**Comment by daira** [comment]
@daira:
> is that fix in its own commit that I can cherry pick?
> ah, yes: https://github.com/zcash/zcash/pull/4060/commits/af48781fb134d05f545a0586cef37a6716f25851
> I would like to include that in 2.0.7
> *tests it*
> yep that fixes it. any objection to including it in the Blossom testnet activation PR I'm writing? 

@str4d:
> Not from me

**Comment by daira** [comment]
Now I'm getting a different error for some of the tests, e.g.
```
=== Running testscript wallet_changeindicator.py ===
Initializing test directory /tmp/testCRkSwY
Unexpected exception caught during testing: [Errno 2] No such file or directory: 'cache/node1'
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/test_framework.py", line 117, in main
    self.setup_chain()
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/test_framework.py", line 36, in setup_chain
    initialize_chain(self.options.tmpdir)
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/util.py", line 146, in initialize_chain
    shutil.copytree(from_dir, to_dir)
  File "/usr/lib/python2.7/shutil.py", line 194, in copytree
    names = os.listdir(src)
Stopping nodes
Traceback (most recent call last):
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/wallet_changeindicator.py", line 79, in <module>
    WalletChangeIndicatorTest().main()
  File "/home/daira/zecc/zcash.pristine/qa/rpc-tests/test_framework/test_framework.py", line 137, in main
    stop_nodes(self.nodes)
AttributeError: 'WalletChangeIndicatorTest' object has no attribute 'nodes'
!!! FAIL: wallet_changeindicator.py !!!
```

**Comment by daira** [comment]
The pertinent error seems to be "No such file or directory: 'cache/node1'". The later AttributeError when cleaning up is likely a red herring (because the cleanup code didn't expect that a failure would occur before `self.nodes` had been set).

**Comment by daira** [comment]
Deleting the `cache` directory fixed this problem. Perhaps the `cache` directory should be automatically deleted by `qa/pull-tester/rpc-tests.sh`.

**Comment by Eirik0** [comment]
This is fixed in #4105
